pkg/controllers/management/clusterupstreamrefresher/cluster_upstream_refresher.go

```diff
@@ -0,0 +1,274 @@
+	wranglerv1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/robfig/cron"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	noKEv2Provider           = "none"
+	lastRefreshTime          = "clusters.management.cattle.io/ke-last-refresh"
+	refreshCronSettingFormat = "%s-refresh-cron"
+)
+
+type clusterRefreshController struct {
+	secretsCache        wranglerv1.SecretCache
+	clusterClient       v3.ClusterClient
+	clusterCache        v3.ClusterCache
+	clusterEnqueueAfter func(name string, duration time.Duration)
+}
+
+// for other cloud drivers, please edit HERE
+type clusterConfig struct {
+	eksConfig *eksv1.EKSClusterConfigSpec
+}
+
+func Register(ctx context.Context, wContext *wrangler.Context) {
+	c := clusterRefreshController{
+		secretsCache:        wContext.Core.Secret().Cache(),
+		clusterClient:       wContext.Mgmt.Cluster(),
+		clusterCache:        wContext.Mgmt.Cluster().Cache(),
+		clusterEnqueueAfter: wContext.Mgmt.Cluster().EnqueueAfter,
+	}
+
+	wContext.Mgmt.Cluster().OnChange(ctx, "eks-operator-controller", c.onClusterChange)
+}
+
+func (c *clusterRefreshController) onClusterChange(key string, cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	if cluster == nil || cluster.DeletionTimestamp != nil {
+		return cluster, nil
+	}
+
+	provider, ready := getProviderAndReadyStatus(cluster)
+	if provider == noKEv2Provider {
+		// not a KEv2 cluster
+		return cluster, nil
+	}
+
+	if !ready {
+		logrus.Infof("initial upstream spec for cluster [%s] has not been set by cluster handler yet, skipping", cluster.Name)
+		return cluster, nil
+	}
+
+	var lastRefreshTime string
+	if cluster.Annotations != nil {
+		lastRefreshTime = cluster.Annotations[lastRefreshTime]
+	}
+
+	providerRefreshInterval, err := getProviderRefreshInterval(provider)
+	if err != nil {
+		return cluster, err
+	}
+
+	nextRefreshTime, err := getNextProviderRefreshTime(providerRefreshInterval, lastRefreshTime)
+	if err != nil {
+		return cluster, err
+	}
+
+	if !time.Now().After(nextRefreshTime) {
+		return cluster, err
+	}
+
+	cluster, err = c.refreshClusterUpstreamSpec(cluster, provider)
+	if err != nil {
+		return cluster, err
+	}
+
+	c.clusterEnqueueAfter(key, providerRefreshInterval)
+
+	return cluster, nil
+}
+
+// getProviderAndReadyStatus returns the managed cluster provider of the given
+// custer and whether it is ready to be refresh and synced.
+func getProviderAndReadyStatus(cluster *mgmtv3.Cluster) (string, bool) {
+	// for other cloud drivers, please edit HERE
+	switch {
+	case cluster.Spec.EKSConfig != nil:
+		if cluster.Status.EKSStatus.UpstreamSpec == nil {
+			logrus.Infof("initial upstream spec for cluster [%s] has not been set by cluster handler yet, skipping", cluster.Name)
+			return apimgmtv3.ClusterDriverEKS, false
+		}
+		return apimgmtv3.ClusterDriverEKS, true
+	default:
+		return noKEv2Provider, false
+	}
+}
+
+// getProviderRefreshInterval returns the duration that should pass between
```

Comment:
As discussed, this should be revamped to use seconds. If cron format is found in old setting we migrate to seconds.